File: src/flutter/test/purchases_test.dart
import 'dart:convert';
import 'package:flutter_test/flutter_test.dart';
import 'package:ansight_flutter/ansight.dart';

void main() {
  test(
      'purchase facade delegates to native core and preserves missing evidence',
      () async {
    final calls = <Map<String, Object?>>[];
    final purchases = PurchaseDiagnostics((method, arguments) async {
      expect(method, 'purchaseCommand');
      calls.add(Map<String, Object?>.from(
          jsonDecode(arguments!['json']! as String) as Map));
      return <String, Object?>{'json': '{"status":"inconclusive"}'};
    });
    await purchases.record(const PurchaseObservation(
        productId: 'premium', source: 'app', entitled: false));
    final report = await purchases.validate(
        productId: 'premium',
        expectedEntitled: true,
        requireBackendVerification: true);
    expect(calls[0]['action'], 'record');
    expect(calls[1]['action'], 'purchases.validate');
    expect(report['status'], 'inconclusive');
  });
}

File: src/react-native/test/purchases.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const { createPurchaseDiagnostics } = require('../purchases');

test('purchase facade forwards typed observations and expectations to native core', async () => {
  const calls = [];
  const purchases = createPurchaseDiagnostics(async (json) => { calls.push(JSON.parse(json)); return '{"success":true}'; });
  await purchases.register();
  await purchases.record({ productId: 'premium', source: 'app', entitled: false });
  await purchases.validate({ productId: 'premium', expectedEntitled: true, requireBackendVerification: true, expectedDeliveryCount: 1 });
  await purchases.refreshStoreKit(['premium']);
  await purchases.clear();
  assert.deepEqual(calls.map(c => c.action), ['register', 'record', 'purchases.validate', 'refreshStoreKit', 'clear']);
  assert.equal(calls[1].observation.entitled, false);
  assert.equal(typeof calls[1].observation.observedAt, 'number');
  assert.equal(calls[2].arguments.requireBackendVerification, true);
});

test('native failures propagate instead of becoming successful empty snapshots', async () => {
  const purchases = createPurchaseDiagnostics(async () => { throw new Error('native unavailable'); });
  await assert.rejects(purchases.snapshot(), /native unavailable/);
});
